### PR TITLE
KOGITO-7612 - [Kogito Guides] Implement gRPC streaming scenarios

### PR DESCRIPTION
**Pull request by MarianMacik**
JIRA: [KOGITO-7612](https://issues.redhat.com/browse/KOGITO-7612)

PR set:
* https://github.com/kiegroup/kogito-docs/pull/87
* https://github.com/kiegroup/kogito-examples/pull/1351

Extended the gRPC guide with streaming scenarios. I tried to kept it concise, not to repeat everything, just pinpoint differences to the simple gRPC scenario which was already described.

### CODE REVIEW

File: serverlessworkflow/modules/ROOT/pages/service-orchestration/orchestration-of-grpc-services.adoc [hunk 1]
```diff
@@ -14,20 +13,30 @@ As a developer, you can invoke a gRPC service using a Serverless Workflow applic
...
+This file defines a greeter service, which consists of sayHello methods, each covering a different gRPC scenario:
+
+1. The `sayHello` method accepts two parameters, including name of the user and an optional language to use in the greeting message (English by default). The `sayHello` method returns the greeting message in the appropriate language.
+2. The `SayHelloAllLanguages` accepts just the name parameter and streams greetings in all languages supported by the service. This represents the server-side streaming scenario.
+3. The `SayHelloMultipleLanguagesAtOnce` accepts a stream of name-language pairs as parameters and returns corresponding greetings as one block of text after the streaming is finished. This is the client-side streaming scenario.
+4. The `SayHelloMultipleLanguages` accepts a stream of name-language pairs and streams back greetings with a respective name and in a respective language.
+5. The `SayHelloMultipleLanguagesError` is the same as before but is used just for testing purposes. It is not relevant for this guide.
```
Comment (fjtirado) [inline serverlessworkflow/modules/ROOT/pages/service-orchestration/orchestration-of-grpc-services.adoc:28]: If it is not relavant for this guide, we can skip mentioning it. 
We are not describing the proto, just mentioning those things in the proto which are relevant to the guide

**Review by fjtirado — APPROVED** [review]
Just a minor comment. 
Great job @MarianMacik

File: serverlessworkflow/modules/ROOT/pages/service-orchestration/orchestration-of-grpc-services.adoc (hunk 1 above)
Comment (MarianMacik) [reply serverlessworkflow/modules/ROOT/pages/service-orchestration/orchestration-of-grpc-services.adoc]: Fixed.

File: serverlessworkflow/modules/ROOT/pages/service-orchestration/orchestration-of-grpc-services.adoc [hunk 2]
```diff
@@ -14,20 +13,29 @@ As a developer, you can invoke a gRPC service using a Serverless Workflow applic
...
 
 Relevant files of this Workflow Application are:
 
-* link:{grpc_example_url}/serverless-workflow-greeting-server-rpc-quarkus/src/main/proto/greeting.proto[gRPC proto file]: 
+* link:{grpc_example_url}/serverless-workflow-greeting-server-rpc-quarkus/src/main/proto/greeting.proto[gRPC proto file]:
 
-This file defines a greeter service, which consists of the `sayHello` method. The `sayHello` method accepts two parameters, including name of the user and an optional language to use in the greeting message (English by default). The `sayHello` method returns the greeting message in the appropriate language.
+This file defines a greeter service, which consists of sayHello methods, each covering a different gRPC scenario:
+
+1. The `sayHello` method accepts two parameters, including name of the user and an optional language to use in the greeting message (English by default). The `sayHello` method returns the greeting message in the appropriate language.
```
Comment (hmanwani-rh) [inline serverlessworkflow/modules/ROOT/pages/service-orchestration/orchestration-of-grpc-services.adoc:24]:
```suggestion
* The `sayHello` method accepts two parameters, including the name of the user and an optional language to use in the greeting message (English by default). The `sayHello` method returns the greeting message in the appropriate language.
```

File: serverlessworkflow/modules/ROOT/pages/service-orchestration/orchestration-of-grpc-services.adoc [hunk 3]
```diff
@@ -14,20 +13,29 @@ As a developer, you can invoke a gRPC service using a Serverless Workflow applic
...
+2. The `SayHelloAllLanguages` accepts just the name parameter and streams greetings in all languages supported by the service. This represents the server-side streaming scenario.
```
Comment (hmanwani-rh) [inline serverlessworkflow/modules/ROOT/pages/service-orchestration/orchestration-of-grpc-services.adoc:25]:
```suggestion
* The `SayHelloAllLanguages` accepts the name parameter and streams greetings in all languages supported by the service. This represents the server-side streaming scenario.
```

File: serverlessworkflow/modules/ROOT/pages/service-orchestration/orchestration-of-grpc-services.adoc [hunk 4]
```diff
@@ -14,20 +13,29 @@ As a developer, you can invoke a gRPC service using a Serverless Workflow applic
...
+3. The `SayHelloMultipleLanguagesAtOnce` accepts a stream of name-language pairs as parameters and returns corresponding greetings as one block of text after the streaming is finished. This is the client-side streaming scenario.
```
Comment (hmanwani-rh) [inline serverlessworkflow/modules/ROOT/pages/service-orchestration/orchestration-of-grpc-services.adoc:26]:
```suggestion
* The `SayHelloMultipleLanguagesAtOnce` accepts a stream of name-language pairs as parameters and returns corresponding greetings as one block of text after the streaming is finished. This is the client-side streaming scenario.
```

File: serverlessworkflow/modules/ROOT/pages/service-orchestration/orchestration-of-grpc-services.adoc [hunk 5]
```diff
@@ -14,20 +13,29 @@ As a developer, you can invoke a gRPC service using a Serverless Workflow applic
...
+4. The `SayHelloMultipleLanguages` accepts a stream of name-language pairs and streams back greetings with a respective name and in a respective language.
```
Comment (hmanwani-rh) [inline serverlessworkflow/modules/ROOT/pages/service-orchestration/orchestration-of-grpc-services.adoc:27]:
```suggestion
* The `SayHelloMultipleLanguages` accepts a stream of name-language pairs and streams back greetings with a respective name and in a respective language.
```

File: serverlessworkflow/modules/ROOT/pages/service-orchestration/orchestration-of-grpc-services.adoc [hunk 6]
```diff
@@ -84,9 +92,32 @@ The execution of the workflow consists of a single link:{spec_doc_url}#operation
 ----
 
 In the previous example, the action invokes the `sayHello` RPC method using two model variables: `name` and `language`. These parameters are provided as part of the REST call that starts the workflow.
-The response of the method execution is link:{grpc_json_mapping}[mapped] to a JSON object and merged with the workflow model. 
+The response of the method execution is link:{grpc_json_mapping}[mapped] to a JSON object and merged with the workflow model.
+
+For the remaining scenarios, have a look at the other workflow definition files:
```
Comment (hmanwani-rh) [inline serverlessworkflow/modules/ROOT/pages/service-orchestration/orchestration-of-grpc-services.adoc:97]:
```suggestion
For the remaining scenarios, see the following workflow definition files:
```

File: serverlessworkflow/modules/ROOT/pages/service-orchestration/orchestration-of-grpc-services.adoc [hunk 7]
```diff
@@ -162,4 +195,56 @@ curl -X POST -H 'Content-Type:application/json' -H 'Accept:application/json' -d
 
 In response, you will see the greeting in Spanish language.
 
+=== Server-side streaming gRPC
+
+[source,shell]
```
Comment (hmanwani-rh) [inline serverlessworkflow/modules/ROOT/pages/service-orchestration/orchestration-of-grpc-services.adoc:200]:
```suggestion
.Example request
[source,shell]
```

File: serverlessworkflow/modules/ROOT/pages/service-orchestration/orchestration-of-grpc-services.adoc [hunk 8]
```diff
@@ -162,4 +195,56 @@ curl -X POST -H 'Content-Type:application/json' -H 'Accept:application/json' -d
...
+----
+curl -X POST -H 'Content-Type:application/json' -H 'Accept:application/json' -d '{"workflowdata" : {"name": "John"}}' http://localhost:8080/jsongreetserverstream
+----
+
+The answer should be similar to this one
```
Comment (hmanwani-rh) [inline serverlessworkflow/modules/ROOT/pages/service-orchestration/orchestration-of-grpc-services.adoc:205]:
```suggestion

```

File: serverlessworkflow/modules/ROOT/pages/service-orchestration/orchestration-of-grpc-services.adoc [hunk 9]
```diff
@@ -162,4 +195,56 @@ curl -X POST -H 'Content-Type:application/json' -H 'Accept:application/json' -d
...
+
+[source,shell]
```
Comment (hmanwani-rh) [inline serverlessworkflow/modules/ROOT/pages/service-orchestration/orchestration-of-grpc-services.adoc:207]:
```suggestion
.Example response
[source,shell]
```

File: serverlessworkflow/modules/ROOT/pages/service-orchestration/orchestration-of-grpc-services.adoc [hunk 10]
```diff
@@ -162,4 +195,56 @@ curl -X POST -H 'Content-Type:application/json' -H 'Accept:application/json' -d
...
+----
+{"id":"665911c5-36ee-40b7-93dd-a2328f969c73","workflowdata":{"name":"John","response":[{"message":"Hello from gRPC service John"},{"message":"Saludos desde gRPC service John"}]}}
+----
+
+=== Client-side streaming gRPC
+
+[source,shell]
```
Comment (hmanwani-rh) [inline serverlessworkflow/modules/ROOT/pages/service-orchestration/orchestration-of-grpc-services.adoc:214]:
```suggestion
.Example request
[source,shell]
```

File: serverlessworkflow/modules/ROOT/pages/service-orchestration/orchestration-of-grpc-services.adoc [hunk 11]
```diff
@@ -162,4 +195,56 @@ curl -X POST -H 'Content-Type:application/json' -H 'Accept:application/json' -d
...
+----
+curl -X POST -H 'Content-Type:application/json' -H 'Accept:application/json' -d '{"workflowdata" : {
+                                                                                                    "helloRequests" : [
+                                                                                                                        {"name" : "Javierito", "language":"Spanish"},
+                                                                                                                        {"name" : "John", "language":"English"},
+                                                                                                                        {"name" : "Jan", "language":"Czech"}
+                                                                                                                      ]}}' http://localhost:8080/jsongreetclientstream
+----
+
+The answer should be similar to this one
```
Comment (hmanwani-rh) [inline serverlessworkflow/modules/ROOT/pages/service-orchestration/orchestration-of-grpc-services.adoc:224]:
```suggestion

```

File: serverlessworkflow/modules/ROOT/pages/service-orchestration/orchestration-of-grpc-services.adoc [hunk 12]
```diff
@@ -162,4 +195,56 @@ curl -X POST -H 'Content-Type:application/json' -H 'Accept:application/json' -d
...
+
+[source,shell]
```
Comment (hmanwani-rh) [inline serverlessworkflow/modules/ROOT/pages/service-orchestration/orchestration-of-grpc-services.adoc:226]:
```suggestion
.Example response
[source,shell]
```

File: serverlessworkflow/modules/ROOT/pages/service-orchestration/orchestration-of-grpc-services.adoc [hunk 13]
```diff
@@ -162,4 +195,56 @@ curl -X POST -H 'Content-Type:application/json' -H 'Accept:application/json' -d
...
+----
+{"id":"abece3f9-0797-4c10-a1f5-8f3929724689","workflowdata":{"helloRequests":[{"name":"Javierito","language":"Spanish"},{"name":"John","language":"English"},{"name":"Jan","language":"Czech"}],"message":"Saludos desde gRPC service Javierito\nHello from gRPC service John\nHello from gRPC service Jan"}}
+----
+
+=== Bidirectional streaming gRPC
+
+[source,shell]
```
Comment (hmanwani-rh) [inline serverlessworkflow/modules/ROOT/pages/service-orchestration/orchestration-of-grpc-services.adoc:233]:
```suggestion
.Example request
[source,shell]
```

File: serverlessworkflow/modules/ROOT/pages/service-orchestration/orchestration-of-grpc-services.adoc [hunk 14]
```diff
@@ -162,4 +195,56 @@ curl -X POST -H 'Content-Type:application/json' -H 'Accept:application/json' -d
...
+----
+curl -X POST -H 'Content-Type:application/json' -H 'Accept:application/json' -d '{"workflowdata" : {
+                                                                                                    "helloRequests" : [
+                                                                                                                        {"name" : "Javierito", "language":"Spanish"},
+                                                                                                                        {"name" : "John", "language":"English"},
+                                                                                                                        {"name" : "Jan", "language":"Czech"}
+                                                                                                                      ]}}' http://localhost:8080/jsongreetbidistream
+----
+
+The answer should be similar to this one
```
Comment (hmanwani-rh) [inline serverlessworkflow/modules/ROOT/pages/service-orchestration/orchestration-of-grpc-services.adoc:243]:
```suggestion

```

File: serverlessworkflow/modules/ROOT/pages/service-orchestration/orchestration-of-grpc-services.adoc [hunk 15]
```diff
@@ -162,4 +195,56 @@ curl -X POST -H 'Content-Type:application/json' -H 'Accept:application/json' -d
...
+
+[source,shell]
```
Comment (hmanwani-rh) [inline serverlessworkflow/modules/ROOT/pages/service-orchestration/orchestration-of-grpc-services.adoc:245]:
```suggestion
.Example response
[source,shell]
```

**Review by hmanwani-rh — COMMENTED** [review]
@MarianMacik Added a few suggestions below. Please take a look and let me know if you've any questions.

**Comment by MarianMacik** [comment]
@hmanwani-rh Thanks for the review, all suggestions have been applied.

**Review by hmanwani-rh — APPROVED** [review]

**Pull request merged**